lms/djangoapps/grades/rest_api/v1/gradebook_views.py

```diff
@@ -1044,44 +1047,60 @@ def get(self, request, subsection_id):
                 developer_message='Invalid UserID',
                 error_code='invalid_user_id'
             )
+        psg_exists = True
+        override = None
+        history = []
 
         try:
             original_grade = PersistentSubsectionGrade.read_grade(user_id, usage_key)
-        except PersistentSubsectionGrade.DoesNotExist:
-            results = SubsectionGradeResponseSerializer({
-                'original_grade': None,
-                'override': None,
-                'history': [],
-                'subsection_id': usage_key,
-                'user_id': user_id,
-                'course_id': None,
-            })
-
-            return Response(results.data)
+            grade_data = {
+                'earned_all': original_grade.earned_all,
+                'possible_all': original_grade.possible_all,
+                'earned_graded': original_grade.earned_graded,
+                'possible_graded': original_grade.possible_graded,
+            }
 
+        except PersistentSubsectionGrade.DoesNotExist:
+            psg_exists = False
+            student = get_user_model().objects.get(id=user_id)
+            course_structure = get_course_blocks(student, usage_key)
+            course = modulestore().get_course(usage_key.course_key)
```

Comment:
Could you explain a little more about the subtree - I assume we need to pass in a course object, and I don't see course_structure having a property

Reply:
Sure, this is pertinent for the initialization of `SubsectionGradeFactory` below.  That thing's `__init__` does this: `self.course_data = course_data or CourseData(student, course=course, structure=course_structure)`.  You could pass `course=None` into `SubsectionGradeFactory.__init__()`, and that `None` would be passed into the `course` parameter for the `CourseData.__init__()` call.

For the `self.course_data` variable inside the factory, we really only need to look at it's `structure` attribute - see the line that does:
```python
subsection_grade = CreateSubsectionGrade(
    subsection, self.course_data.structure, self._submissions_scores, self._csm_scores,
)
```

If you look at the `CourseData.structure` property, you'll see
```python
    @property
    def structure(self):
        if self._structure is None:
            self._structure = get_course_blocks(
                self.user,
                self.location,
                collected_block_structure=self._collected_block_structure,
            )
        return self._structure
```

You're passing in a non-null `structure` to the SubsectionGradeFactory, which gets passed into the CourseData object and stored in the `_structure` attribute, so you basically already did all of the hard work that `structure` does - it will just return `self._structure`.

So in that whole path from the `get()` view function to `CreateSubsectionGrade(...)`, we didn't need a `course` object anywhere - all we needed was the structure created by `course_structure = get_course_blocks(student, usage_key)`.  We didn't need to fetch the course from the modulestore, just the subtree for the subsection located at `usage_key`.